### ISSUE TITLE
ES container silently fails when running on low resources

### ISSUE THREAD

**Issue by sebgl** [elastic/cloud-on-k8s#1097]
How to reproduce:

* setup a Minikube VM with 2 CPUs and 2GB RAM
* deploy the operator on Minikube (`make deploy`)
* deploy the sample stack (`make samples`)

After a few seconds, the ES container crashes with the following log:

```
{"level":"info","ts":1560785047.2425053,"logger":"process-manager","msg":"Update process state","action":"initialization","id":"es","state":"failed","pid":0}
{"level":"info","ts":1560785047.2429287,"logger":"process-manager","msg":"Starting..."}
{"level":"info","ts":1560785047.3111901,"logger":"process-manager","msg":"Update process state","action":"start","id":"es","state":"started","pid":12}
{"level":"info","ts":1560785047.3115153,"logger":"process-manager","msg":"Started"}
{"level":"info","ts":1560785047.3403187,"logger":"keystore-updater","msg":"Waiting for Elasticsearch to be ready"}
OpenJDK 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
{"level":"info","ts":1560785056.5088394,"logger":"process-manager","msg":"Update process state","action":"terminate","id":"es","state":"failed","pid":12}
{"level":"info","ts":1560785056.543558,"logger":"process-manager","msg":"Exit","reason":"process failed","code":-1
```

It seems to be caused by low resources (not enough RAM), but there is no clear indication that is the case. With https://github.com/elastic/cloud-on-k8s/pull/1029 we removed the default resources if not provided by the user.

Should we re-introduce a 1GB memory **requests** by default? At least when running on Minikube with low resources we get a clearer error (pod stays pending with an explicit message displayed on `kubectl describe`).

**Comment by barkbay** [comment]
If by default ES starts with a Xm(s|x) of 1Gi then a request of 1Gi is not enough.
If the container is killed by the OOMKiller it should be reported through some events and in the status, not sure we can do more about it. 

Not sure also about the default request since it may be set by a `LimitRange`, maybe we should add the request in our examples with a comment/documentation ? (c.f. https://github.com/elastic/cloud-on-k8s/pull/1029#pullrequestreview-247393459 on a similar topic)

**Comment by sebgl** [comment]
> If the container is killed by the OOMKiller it should be reported through some events and in the status, not sure we can do more about it.

I think that's not what happens here (but I did not look at it close enough): the container is not killed, it looks like the ES process simply cannot start, so the process-manager exits.

I'm +1 on having most samples explicitly specify memory requirements, but it does ruin a bit the quickstart experience.
If we're not specifying any minimum memory request, can we maybe do something at the process-manager level to give more insights in the logs about why ES is not starting?

**Comment by thbkrkr** [comment]
I did a test: in the same condition (in minikube), I started the ES process without the process-manager, inside the ES container. I confirm that the exit code of the Elasticsearch process is `137`.

```bash
$ docker run --rm -ti docker.elastic.co/elasticsearch/elasticsearch:7.1.0 bash
[root@4755eeccde33 elasticsearch]# /usr/local/bin/docker-entrypoint.sh; echo $?
OpenJDK 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
Killed
137
```

So that would indicate that the process manager does not forward well the ES exit code.

**Comment by thbkrkr** [comment]
By default, the [`docker-entrypoint.sh`](https://github.com/elastic/elasticsearch/blob/a2bdfb9041528e52ccc21369886f811802d12ce6/distribution/docker/src/docker/bin/docker-entrypoint.sh) and the script [`bin/elasticsearch`](https://github.com/elastic/elasticsearch/blob/a2bdfb9041528e52ccc21369886f811802d12ce6/distribution/src/bin/elasticsearch) use `exec` so that the final Elasticsearch process becomes the container's PID 1.
In our case, the Elasticsearch process must stay a child process to let the parent process, the process manager, be responsible for handling and forwarding signals as appropriate.

In other words, removing the `exec` from one of the two scripts, it works.

**Comment by sebgl** [comment]
We discussed this problem again with @nkvoll and concluded that we should probably set a default resources memory request of 2G (2x default ES JVM of 1Gi).
If we don't, people running the quickstart are able to deploy 30 ES nodes on a single minikube host: all pods are running but will randomly fail one after the other. This looks quite bad. Mostly, it *doesn't make sense* for users to run the default Elasticsearch (1G heap size) without at least 2G resources request. Also, it doesn't make sense to have no memory request at all on the ES container. (Not having resource limits does make sense though). We still want people to have some control over this default though: being able to override it with their own, or being able to explicitly disable our default to use their cluster limit range.

No resource specified, apply a default memory request of 2Gi (no memory limit):
```yaml
apiVersion: elasticsearch.k8s.elastic.co/v1alpha1
kind: Elasticsearch
metadata:
  name: quickstart
spec:
  version: 7.1.0
  nodes:
  - nodeCount: 1
```

User sets their own requests and limits (to be documented):
```yaml
apiVersion: elasticsearch.k8s.elastic.co/v1alpha1
kind: Elasticsearch
metadata:
  name: quickstart
spec:
  version: 7.1.0
  nodes:
  - nodeCount: 1
    podTemplate:
      spec:
        containers:
        - name: elasticsearch
          env:
          - name: ES_JAVA_OPTS
             value: "-Xms2G -Xmx2G" 
          resources:
            requests:
              memory: 4Gi
            limits:
              memory: 4Gi   
```

User wants to default to the cluster limit range, specify an empty resource requests (!= nil):
```yaml
apiVersion: elasticsearch.k8s.elastic.co/v1alpha1
kind: Elasticsearch
metadata:
  name: quickstart
spec:
  version: 7.1.0
  nodes:
  - nodeCount: 1
    podTemplate:
      spec:
        containers:
        - name: elasticsearch
          resources:
            requests: {}
```

**Comment by barkbay** [comment]
What about creating a [LimitRequest](https://kubernetes.io/docs/concepts/policy/limit-range/) in the managed namespace?
This is what is supposed to set some constraints on deployment in K8S.

**Comment by barkbay** [comment]
Nevermind, it would apply to any pod in the namespace (including Kibana).
